automation/functions/delivery.py

```diff
@@ -0,0 +1,174 @@
+from python_http_client.exceptions import BadRequestsError
+from requests.exceptions import HTTPError
+from sendgrid.helpers.mail import Mail
+import structlog
+from structlog.contextvars import bind_contextvars
+
+from automation.utils.templates import render
+
+log = structlog.get_logger("send_delivery_email")
+
+# Turn this off for debugging
+SEND_MAIL = True
+
+
+def on_assigned(
+    ticket,
+    *,
+    member_table,
+    inventory,
+    sendgrid_client,
+    from_email,
+    reply_to=None,
+):
+    bind_contextvars(ticket_id=ticket.ticket_id)
+    log.info("Sending delivery email")
+
+    problem = check_ready_to_send(ticket)
+    if problem:
+        msg = "Ticket not ready to send"
+        log.error(msg, problem=problem)
+        return
+
+    delivery_volunteers = get_delivery_volunteers(ticket, member_table)
+    email = render_email_template(ticket, delivery_volunteers, inventory)
+    email.from_email = from_email
+    if reply_to:
+        email.add_cc(reply_to)
+        email.reply_to = reply_to
+
+    if not SEND_MAIL:
+        log.info(
+            "Would send email", to=email.get()["personalizations"][0]["to"]
+        )
+        return
+    try:
+        sendgrid_client.send(email)
+        log.info("Sent email")
+    except BadRequestsError as e:
+        msg = f"Error sending email: {e.status_code} {e.reason}"
+        log.error(msg, body=e.body, exc_info=True)
+        raise DeliveryEmailError(msg)
+
+
+def check_ready_to_send(ticket):
```

Comment:
Any of these cases will cause the poll function to stop processing that ticket and move on. The intended workflow is the delivery volunteer would say "I didn't get an email", the coordinator would ask automation what happened, and we'd go look at logs to see one of the error messages below about that ticket id.

Reply:
I like that idea a lot. Sort of want to save it for a follow up though, especially because it's a schema change.

Reply:
!59